### PR TITLE
Add support for `Union` types, including pipe syntax

### LINKED ISSUE #520: Allow parsing of complex union types
Currently, any complex union (anything except `| None`) fails to parse and is excluded from the generated code. The signature below is an example:

```python
def foo(a: str | int | float) -> int:
   ... 
```

This API is common for `AnyStr` (`str | bytes`):

e.g.

```python

def encode(f: str | bytes) -> bytes:
  ...

# or file paths
def open_file(path: str | pathlib.Path) -> fp:
   ...
```

Suggest we implement this in two stages. The first being more urgent

1. Allow parsing of these signatures and reflect to `Any` i.e. `PyObject`
2. Generate overloads for each combination of types (within reason)

Related: #23

### PR DESCRIPTION
**Pull request by atifaziz**
This PR closes #520. It significantly improves the Python type parser by implementing comprehensive support for Union types, including both the traditional `Union[...]` syntax and modern pipe syntax (`A | B`). The changes introduce type normalization, de-duplication, and optimization strategies.

It should help with efforts like #499.

The implementation is optimized for up to 3 unions and then takes a slower path for more esoteric expressions of a Union that requires normalization of sorts.

## Design Changes

The core design philosophy shifts from treating Union types as edge cases to handling them as first-class citizens in the type system.

1. **Unified Type Parsing**: Consolidates the handling of different type syntaxes (None, Union, pipe syntax) into a single, coherent parsing strategy
2. **Normalization**: Automatically optimizes Union types by flattening nested unions, deduplicating types, and converting common patterns to their canonical forms
3. **Equality and Hashing**: Implements proper value equality and hash code caching for `PythonTypeSpec` to support de-duplication and performance optimization

### Union Type Normalization

The new `Union` static method implements sophisticated normalization rules:

- **Single Type Optimization**: `Union[int]` → `int`
- **Optional Type Recognition**: `Union[int, None]` → `Optional[int]`
- **Nested Union Flattening**: `Union[Union[int, str], bool]` → `Union[int, str, bool]`
- **Deduplication**: `Union[int, int, str]` → `Union[int, str]`
- **Performance Optimization**: Uses HashSet for large unions (>8 types) to improve containment checking

### Equality and Hashing Implementation

- Two `PythonTypeSpec` instances are considered equal if they have the same name and arguments
- `PythonTypeSpec` hash code computation is cached since the type is fully immutable

### Test Coverage Expansion

The test suite has been significantly expanded to cover:

- **Union Normalization**: 20+ test cases covering various union scenarios
- **Pipe Syntax Support**: Tests for the modern `A | B` syntax
- **Edge Cases**: Complex nested unions, large unions with duplicates, mixed syntax patterns

## Breaking Changes

This is a **non-breaking change** for end users.

The parser now accepts more valid Python type syntax and produces more normalized output, but all previously supported syntax continues to work.

## Limitations and Further Work

Qualified forms like `typing.Union` and `typing.Optional` are not currently recognised, but this can be addressed in the future through a better re-modeling of `PythonTypeSpec` so that it can also be addressed for `Literal`.

### CODE REVIEW

File: src/CSnakes.SourceGeneration/Parser/Types/PythonTypeSpec.cs [hunk 1]
```diff
@@ -18,4 +21,96 @@ public override string ToString() =>
...
+            // Optimise for common cases of Union with two...
+            case [{ Name: not "Union" and not "Optional" } a, { Name: not "Union" and not "Optional" } b]
+                when a != b: return new("Union", [a, b]);
+            // ...or three different types
+            case [{ Name: not "Union" and not "Optional" } a, { Name: not "Union" and not "Optional" } b, { Name: not "Union" and not "Optional" } c]
+                when a != b && a != c && b != c:
+                return new("Union", [a, b, c]);
+            default:
+            {
+                var list = ImmutableArray.CreateBuilder<PythonTypeSpec>(); // unique in order of appearance
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply src/CSnakes.SourceGeneration/Parser/Types/PythonTypeSpec.cs:47]: List is intentional. 😁

File: src/CSnakes.SourceGeneration/Parser/PythonParser.TypeDef.cs [hunk 2]
```diff
@@ -58,38 +57,27 @@ select ImmutableArray.CreateRange(callable.Parameters.Append(callable.Return)),
...
-
-        // Parser for one of the above or "None"
+            select type;
 
-        return Parse.OneOf(noneUnionParser.Try(),
-                           unionNoneParser.Try(),
-                           Token.EqualTo(PythonToken.None)
-                                .IgnoreThen(Parse.Return<PythonToken, PythonTypeSpec>(PythonTypeSpec.None)),
-                           someParser);
+        return from ts in parser.AtLeastOnceDelimitedBy(Token.EqualTo(PythonToken.Pipe))
```
Comment (tonybaloney) [inline src/CSnakes.SourceGeneration/Parser/PythonParser.TypeDef.cs:75]: Great, this is cleaner than what we had before

**Review by tonybaloney — APPROVED** [review]
